poptorch_experimental_addons/cpp/distance_matrix.cpp

```diff
@@ -0,0 +1,498 @@
+    }
+
+    void grow(poplar::program::Sequence& prog) const final {
+        auto op = getOp<L1DistanceOp>();
+
+        poplar::Tensor A = getInTensor(0);
+        poplar::Tensor B = getInTensor(1);
+        poplar::Tensor out = l1distance(graph(), A, B, prog, "l1distance");
+        setOutTensor(0, out);
+    }
+};
+
+class L1DistanceGradOpx : public popart::popx::Opx {
+   public:
+    L1DistanceGradOpx(popart::Op* op, popart::popx::Devicex* devicex)
+        : popart::popx::Opx(op, devicex) {
+        verifyOp<L1DistanceGradOp>(op, {CustomGradOperators::L1DistanceGradId});
+    }
+
+    void grow(poplar::program::Sequence& prog) const final {
+        auto op = getOp<L1DistanceGradOp>();
+
+        poplar::Tensor grad = getInTensor(0);
+        poplar::Tensor A = getInTensor(1);
+        poplar::Tensor B = getInTensor(2);
+        auto gradA = l1distancegrad(graph(), A, B, grad, prog, "l1distanceGrad");
+        auto gradB = l1distancegrad(graph(), B, A, grad.transpose(), prog, "l1distanceGrad");
+        setOutTensor(0, gradA);
+        setOutTensor(1, gradB);
+    }
+};
+
+class L2DistanceOpx : public popart::popx::Opx {
+   public:
+    L2DistanceOpx(popart::Op* op, popart::popx::Devicex* devicex) : popart::popx::Opx(op, devicex) {
+        verifyOp<L2DistanceOp>(op, {CustomOperators::L2DistanceId});
+        // add codelets to graph
+        graph().addCodelets(ROOT_PATH
+                            "/poptorch_experimental_addons/cpp/distance_matrix_codelet.cpp");
+    }
+
+    void grow(poplar::program::Sequence& prog) const final {
+        auto op = getOp<L2DistanceOp>();
+
+        poplar::Tensor A = getInTensor(0);
+        poplar::Tensor B = getInTensor(1);
+        poplar::Tensor out = l2distance(graph(), A, B, prog, "l2distance");
+        setOutTensor(0, out);
+    }
+};
+
+class L2DistanceGradOpx : public popart::popx::Opx {
+   public:
+    L2DistanceGradOpx(popart::Op* op, popart::popx::Devicex* devicex)
+        : popart::popx::Opx(op, devicex) {
+        verifyOp<L2DistanceGradOp>(op, {CustomGradOperators::L2DistanceGradId});
+    }
+
+    void grow(poplar::program::Sequence& prog) const final {
+        auto op = getOp<L2DistanceGradOp>();
+
+        poplar::Tensor grad = getInTensor(0);
+        poplar::Tensor A = getInTensor(1);
+        poplar::Tensor B = getInTensor(2);
+        poplar::Tensor dist = getInTensor(3);
+        auto gradA = l2distancegrad(graph(), A, B, dist, grad, prog, "l2distanceGrad");
+        auto gradB = l2distancegrad(graph(), B, A, dist.transpose(), grad.transpose(), prog,
+                                    "l2distanceGrad");
+        setOutTensor(0, gradA);
+        setOutTensor(1, gradB);
+    }
+};
+
+L1DistanceGradOp::L1DistanceGradOp(const L1DistanceOp& fwdOp)
+    : popart::Op(CustomGradOperators::L1DistanceGradId, fwdOp.settings) {}
+
+const std::vector<popart::GradInOutMapper>& L1DistanceGradOp::gradInputInfo() const {
```

Comment:
Maybe these `gradInputInfo` and `gradOutToNonGradIn` functions could be defined inline in the class(?) I can't see them depending on other stuff.

I guess the constructors can't be inlined because of `fwdOp.settings`.